File: Java/Medium/Accepted/shortest-bridge_1.java
class Solution {
    public int shortestBridge(int[][] grid) {
        
        Queue<int[]> q = new LinkedList<>();
        boolean found = false;
        for (int i = 0; i < grid.length; i++) {
            if (found) break;
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    dfs(grid, i, j, q);
                    found = true;
                    break;
                }
            }
        }
        int step = 0;
        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                int[] cur = q.poll();
                for (int[] dir : dirs) {
                    int x = cur[0] + dir[0];
                    int y = cur[1] + dir[1];
                    if (x < 0 || x >= grid.length || y < 0 || y >= grid[0].length || grid[x][y] == 2) continue;
                    if (grid[x][y] == 1) return step;
                    grid[x][y] = 2;
                    q.offer(new int[]{x, y});
                }
            }
            step++;
        }
        return -1;
    }
    
    private void dfs(int[][] grid, int i, int j, Queue<int[]> q) {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || grid[i][j] != 1) return;
        grid[i][j] = 2;
        q.offer(new int[]{i, j});
        dfs(grid, i + 1, j, q);
        dfs(grid, i - 1, j, q);
        dfs(grid, i, j + 1, q);
        dfs(grid, i, j - 1, q);
    }
}

File: C++/Medium/Accepted/shortest-bridge_1.cpp
class Solution {
public:
    int shortestBridge(vector<vector<int>>& grid) {
        
        int n = grid.size();
        int m = grid[0].size();
        vector<pair<int, int>> island1;
        vector<pair<int, int>> island2;
        bool found = false;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) {
                    if (!found) {
                        dfs(grid, island1, i, j);
                    } else {
                        dfs(grid, island2, i, j);
                    }
                    found = true;
                }
            }
        }
        int ans = INT_MAX;
        for (auto& p1 : island1) {
            for (auto& p2 : island2) {
                ans = min(ans, abs(p1.first - p2.first) + abs(p1.second - p2.second) - 1);
            }
        }
        return ans;
    }
    void dfs(vector<vector<int>>& grid, vector<pair<int, int>>& island, int x, int y) {
        if (x < 0 || y < 0 || x >= grid.size() || y >= grid[0].size() || grid[x][y] == 0) {
            return;
        }
        island.push_back({x, y});
        grid[x][y] = 0;
        dfs(grid, island, x + 1, y);
        dfs(grid, island, x - 1, y);
        dfs(grid, island, x, y + 1);
        dfs(grid, island, x, y - 1);
    }
};